exercise-book/src/calculator.md

```diff
@@ -0,0 +1,700 @@
+<!-- markdownlint-disable MD033 -->
+# Calculator
+
+In this exercise we will implement a Reverse-Polish-Notation Calculator library.
+
+You will learn:
+
+* How to write libraries in Rust
+* How to use tests to check library code
+* How to model tree-like data structures in Rust using `enum`
+* How to use `enum` types to model custom errors
+* How to use turbofish syntax to assist type resolution
+* How to handle simple text parsing in Rust
+
+## Task
+
+Write a library that parses and evaluates math expressions that use [Reverse-Polish (postfix) notation](https://en.wikipedia.org/wiki/Reverse_Polish_notation).
+For example, a string `"3 1 + 2 /"` in this notation is equivalent to `(3 + 1) / 2` but unlike the later it does not force us to handle operator precedence using parenthesis.
+
+We will support the 4 basic math operations (`+`, `-`, `*`, `/`) that all will expect two operants, and a `sqr` operation to square a single number.
+
+Here's a basic grammar of expressions that we can expect:
+
+```rust ignore
+expr =
+    number
+    | expr expr '+'
+    | expr expr '-'
+    | expr expr '*'
+    | expr expr '/'
+    | expr 'sqr'
+```
+
+and here are some examples for you:
+
+| Postfix Notation | Traditional notation | Value |
+|------------------|----------------------|-------|
+| 42               | 42                   | 42    |
+| 40 2 +           | 40 + 2               | 42    |
+| 1 3 + 2 /        | (1 + 3) / 2          | 2     |
+
+**Hint:** you can use these examples as basis for unit tests.
+
+Our library should expose a pair of functions:
+
+```rust ignore
+// Takes a string and produces an expression tree
+pub fn parse(input: &str) -> Result<Expr, ParseError>
+
+// Evaluates the expression
+pub fn eval(expr: &Expr) -> Result<i64, ParseError>
+```
+
+### `Expr` type
+
+The `Expr` type has to represent any math expression we can encounter, from a simple number to a complex nested expressions.
+We will use an `enum` type for different expression variants.
+
+Note how we use `Expr` inside a `Square` variant.
+This way we can represent both simple expressions (like `5 sqr`) and something very complex (like `5 sqr sqr sqr` or `3 2 + sqr`).
+
+```rust ignore
+enum Expr {
+    Number(i64),
+    Square(Box<Expr>),
+    // ...
+}
+
+// Here's how a `Square` variant can be created:
+let four = Expr::Square(Box::new(Expr::Number(2)));
+```
+
+### Here's a bit of code to get you started
+
+```rust ignore
+pub enum Expr {}
+
+pub enum ParseError {}
+
+pub fn parse(input: &str) -> Result<Expr, ParseError> {
+    let mut stack: Vec<Expr> = Vec::new();
+    for word in input.split_ascii_whitespace() {
+
+    };
+    assert_eq!(stack.len(), 1);
```

Comment:
I'm not sure that suggesting `unwrap()` by default is the best solution here, but I'll go with it if you're sure it's best as I haven't done this exercise.

Reply:
This actually is a leftover from old times. I've put the starter code as is, and the exercise Help section talks about this `assert_eq!` and the `unwrap` are problematic.

The participants run into it often, and some get confused how their tests are visibly failing, but the error message is not what they have in the test body and instead is something completely unrelated. I turn it into a teachable moment instead, and with this exercise rewrite one of the steps in the solution rewrite this portion of code, too.